File: tree/199.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        
        #for each layer, return right most node
        #use breath first search
        
        #time: O(n)
        #space: O(n)
        
        if not root:
            return root
        
        result = []
        stack = [root]
        
        while stack:
            
            children = []
            while stack:
                right_node = stack.pop()
                children.append(right_node)
            
            result.append(right_node.val)
                
            for i in range(-1, -len(children) - 1, -1):
                if children[i].right:
                    stack.append(children[i].right)
                if children[i].left:
                    stack.append(children[i].left)
        
        return result
        

File: grind75/Jan03/199.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rightSideView(root *TreeNode) []int {
    ret := []int{}
    if root == nil {
        return ret
    }
    q := []*TreeNode{root}
    level := 0
    for len(q) > 0 {
        cq := []*TreeNode{}
        lq := len(q)
        for lq > 0 {
            node := q[0]
            q = q[1:]
            lq--
            if len(ret) == level {
                ret = append(ret, node.Val)
            }
            if node.Right != nil {
                cq = append(cq, node.Right)
            }
            if node.Left != nil {
                cq = append(cq, node.Left)
            }
        }
        q = cq
        level += 1
    }
    return ret
}
